Fix 3D rotation precession

scottshambaugh opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary

https://github.com/matplotlib/matplotlib/pull/28290 introduced "arcball" style rotation to 3D plots, which IMO is a significant enhancement over the previous rotation style. However, there was a change to the original arcball implementation to slow down the speed of rotation, which had the inadvertent side effect that it caused the plot to precess when moving the mouse around. This removed the desirable property that returning the mouse to the same point on the screen restored the original view. See the comments at the end of that issue for some example videos.

This tweaks the math to have the best of both worlds - a 1:1 mapping between screen position and view angle during a single mouse click, and also scaling down the "double angle" high speed rotations to a "single angle" speed.

@MischaMegens2 could you please check the math and weigh in on if you think this makes sense to change for the default?

I think that since this behavior is new for 3.10, it makes sense to prioritize getting it in. The what's new note shouldn't have to change.

See also some more discussion in https://github.com/matplotlib/matplotlib/issues/28408.

Original arcball paper: https://graphicsinterface.org/wp-content/uploads/gi1992-18.pdf

## PR checklist

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

scottshambaugh commented:
Before, showing that returning to the original mouse position has a view that has precessed:

https://github.com/user-attachments/assets/69d87519-0c1c-4fd4-8928-725964aa98a6

After, showing that returning to the original mouse position recovers the original view:

https://github.com/user-attachments/assets/a6dd239f-add8-42f7-b0db-09e5995ec6d8

MischaMegens2 commented:
> @MischaMegens2 could you please check the math and weigh in on if you think this makes sense to change for the default?
> 
I took a quick look at it and I don't think it is proper. Both proposed modifications miss the mark.
The first one is (additions in bold):
**`    scale = np.sqrt(2)/2  # slow down the rate of rotation`**
`    current_vec = self._arcball(self._sx`**`*scale`**`/w, self._sy`**`*scale`**`/h)`
`    new_vec = self._arcball(x`**`*scale`**`/w, y`**`*scale`**`/h)`
The stated purpose is to 'slow down the rate of rotation'. That is not what it does - it shrinks the arcball instead.
(Incidentally, the arcball size is scaled inside the `_arcball()` method already: the first two lines of it read:
`    x *= 2; y *= 2`
I don't see why you would want to introduce the extra `scale`; if the arcball scaling is not to your liking, why not modify it there...?)
The arcball size on screen leaves the rotation unchanged (you'd have to click closer to the center of the figure since the arcball has shrunk, but you will then achieve exactly the identical effect as before.) So the proposed change does not do what it is supposed to do.

The second proposed modification is:
`    q = dq` **`* dq`**` * q`
i.e., the incremental rotation is now performed _twice_. This exacerbates the '"double angle" high speed rotations': instead of achieving a '"single angle" speed', with the proposed PR, we now have it quadrupled... worse! I don't think anyone is looking forward to that.

As for the movie clips, purporting to show that returning to the original mouse position now recovers the original view: you might well be deceived by the shrinking of the arcball - in the clips, the mouse is already somewhat near the edge, and with the shrinking in the 2nd clip, you probably ended up outside the arcball proper, on its border, and consequently, rotations are around a fixed axis (pointing towards you). Of course you return to the original view then.

Would you not rather wait for https://github.com/matplotlib/matplotlib/issues/28408#issuecomment-2190745981, instead of trying to make changes haphazardly in a rush, without really understanding what you are doing?

MischaMegens2 commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -4023,9 +4023,10 @@ def as_cardan_angles(self):
+        qx, qy, qz = self.vector[..., :] / n
         azim = np.arctan2(2*(-qw*qz+qx*qy), qw*qw+qx*qx-qy*qy-qz*qz)
-        elev = np.arcsin( 2*( qw*qy+qz*qx)/(qw*qw+qx*qx+qy*qy+qz*qz))  # noqa E201
+        elev = np.arcsin(np.clip(2*(qx*qz + qw*qy), -1, 1))
```
But why (the np.clip)? I have a hard time imagining how you'd possibly ever get the arcsin argument out of range...

scottshambaugh replied:
There were floating point round-off errors on the macos github CI runners, which was resulting in values inside the expression of (eg) 1.000000001 and leading to an error being thrown here. I'm not sure of a cleaner robust way to do this than clipping after norming above, but am open to suggestions.

MischaMegens2 replied:
All right, touché. I apologize for being so curt yesterday (it was not a good day for me). Let's see if we can make it work as good as we can.
Is there any chance we can extract the qw, qx, qy, and qz from the macos runners where the error shows up? And some of the intermediate results? Hopefully that will give some insight as to why we're confronted with this. (What hardware is used to do the calculations, what processor?)

scottshambaugh replied:
Floating point inaccuracies are something that pop up fairly frequently for matplotlib's tests, and unfortunately we won't ever get perfectly reproducible results across architectures. It's why we use `assert_allclose` a lot of places in testing. That doesn't work in cases like this where you are close to a discrete threshold however, hence the clipping. The good news is that in 99% of cases for matplotlib, the end result is visual so floating point inaccuracies don't matter to the end user.

MischaMegens2 replied:
Well I understand the expediency of adding an `np.clip()` and moving on, it gets this job done; but I thought we had the [IEEE 754 Standard for Floating-Point Arithmetic](https://en.wikipedia.org/wiki/IEEE_754) for quite a while now, to help make things like this reliable and portable. And it is not like we're doing something arcane: just addition, multiplication and division, and not even a whole lot of it. So I'm somewhat surprised by the incompatibility (why only on macos?). Seems you're on to something.

MischaMegens2 commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -4023,9 +4023,10 @@ def as_cardan_angles(self):
-        qx, qy, qz = self.vector[..., :]
+        n = np.sqrt(self.norm)
+        qw = self.scalar / n
+        qx, qy, qz = self.vector[..., :] / n
```
This replaced a single division by 4 of them, and added an extra sqrt to boot. And it made it harder to understand what the range of the arcsin argument is... Improvement?

scottshambaugh replied:
As the quaternion is not normed upon initialization, this step is necessary to have the trig functions following to be valid. In its current use case the quaternions are always unit length, but there is no guarantee of that. I think the alternative to norming here would be to throw an error if this method is called with a non-unit-norm quaternion.

MischaMegens2 replied:
The step is not necessary, the routine is not sensitive to the magnitude of the quaternion, the way it was before (only  ratios occur in the arcsin arguments); the argument of the arcsin is always in range [-1,+1], it seems perfectly valid. Please revert (or provide a counter example).

scottshambaugh replied:
As a counterexample, `qw, qx, qy, qz = 1, 0, 1, 0`. Before the change this would result in elev = 30 deg, and after it results in 45 deg as expected. The sqrt is necessary for norming.

scottshambaugh replied:
Actually, it wasn't immediately obvious to me but I see now that you put in simplified forms that baked the normalization into the trig functions. I can revert this portion of the change.

MischaMegens2 replied:
> Actually, it wasn't immediately obvious to me but I see now that you put in simplified forms that baked the normalization into the trig functions. 

Should we put a note to clarify/make it more obvious, for our future audiences? 

> I can revert this portion of the change.

Yes please :)

scottshambaugh replied:
Done for both

MischaMegens2 reviewed (commented):
I don't really see the benefit.

timhoffm commented:
@MischaMegens2 thanks for the review.

> Would you not rather wait for [#28408 (comment)](https://github.com/matplotlib/matplotlib/issues/28408#issuecomment-2190745981), instead of trying to make changes haphazardly in a rush, without really understanding what you are doing?

In my ears, this comes across a bit harsh. I'm sure this PR was well intended. 3.10 will be released in October and it would be great to have the rotation fixed there. Asking you on the status per your above comment would have been the right first step. But I have to admit, I myself had forgotten your comment on planning to fix this.

So to get things straight: We'd very much welcome if you could do the fix. Is there anything we can help you with to move forward? - You said you wanted to wait for https://github.com/matplotlib/matplotlib/pull/28395. My interpretation of the status is that we're waiting on your rewrite there. But given that it has become documentation-only, can we decouple this from rotation fix. I'd prioritize the fix higher.

scottshambaugh commented:
Hi @MischaMegens2, addressing the points you have brought up:

> The second proposed modification is:
>       `q = dq * dq * q`
> i.e., the incremental rotation is now performed twice. 

This is on purpose, to restore the original Shoemake arcball recipe by undoing the "sqrt" that was done to the rotation quaternion as you explained here https://github.com/matplotlib/matplotlib/pull/28290#issuecomment-2168535036 (quoted below). This could be implemented in other ways but is necessary for the fixed behavior.

> As for the math of it: the 'arcball' as implemented does not strictly follow the arcball recipe of Ken Shoemake; Ken's arcball uses vectors on the (unit) sphere to directly define a rotation quaternion q = v1 v2', and then rotates the figure using v_rotated = q v_original q'. As a consequence, it rotates by twice the angle subtended on the sphere. (This is particurlarly noticeable when adjusting the 'roll', near the edge.) In contrast, the matplotlib arcball takes a square root: q = sqrt(v1 v2'), avoiding the double-speed rotation.

> The first one is (additions in bold):
>       `scale = np.sqrt(2)/2  # slow down the rate of rotation`
>       `current_vec = self._arcball(self._sx*scale/w, self._sy*scale/h)`
>       `new_vec = self._arcball(x*scale/w, y*scale/h)`
> The stated purpose is to 'slow down the rate of rotation'. That is not what it does - it shrinks the arcball instead.

This scaling makes it so that when moving your mouse from the origin in pure x or y directions, the view angle change exactly matches the original behavior you added with the "sqrt" of the rotation quaternion. You can see this in the unchanged tests. This was done with the purpose of "avoiding the double-speed rotation" you called out. The `sqrt(2)/2` is squared to `1/2` inside the `_arcball` function, which is where the doubling factor comes in.

> (Incidentally, the arcball size is scaled inside the _arcball() method already: the first two lines of it read:
>       `x *= 2; y *= 2`
> I don't see why you would want to introduce the extra scale; if the arcball scaling is not to your liking, why not modify it there...?)

I was trying to keep the math inside `_arcball` matching what's in the Shoemake paper (screenshot below). Our radius of 0.5 results in the factor of 2 in the code. IMO it makes more sense to pull the scaling out of the function, but yes it could be implemented differently. I would rather not remove the scale and change the factors inside the function to `sqrt(2)`, as I think the legibility of the functions is more important than saving multiplications.

![image](https://github.com/user-attachments/assets/4dca20ae-d625-4ff8-a41a-69ba1746f9ec)

> As for the movie clips, purporting to show that returning to the original mouse position now recovers the original view: you might well be deceived by the shrinking of the arcball - in the clips, the mouse is already somewhat near the edge, and with the shrinking in the 2nd clip, you probably ended up outside the arcball proper, on its border, and consequently, rotations are around a fixed axis (pointing towards you). Of course you return to the original view then.

Nope, this works whatever your starting position and however you move your mouse. I encourage you to try it out locally!

> Would you not rather wait for https://github.com/matplotlib/matplotlib/issues/28408#issuecomment-2190745981

My understanding of the proposed change was that it was to add the rcparam, not necessarily to fix the default rotation behavior. With 3.10 releasing soon and this new rotation mode going out to users for the first time, I thought a small fix may be appropriate since larger changes can take a while to review. But it's not set in stone, and getting your status and opinion is why I asked for your feedback. If that larger MR is ready I'd be happy to review it and we can discuss whether it makes sense to have this smaller one as a stepping stone at all.

> instead of trying to make changes haphazardly in a rush, without really understanding what you are doing?

We expect all participants, regardless of their experience or position, to engage respectfully and assume good faith. Please focus on the technical merits of the proposal rather than making personal accusations. All contributors to matplotlib are expected to abide by our [code of conduct](https://matplotlib.org/stable/project/code_of_conduct.html).

MischaMegens2 commented:
> So to get things straight: We'd very much welcome if you could do the fix. Is there anything we can help you with to move forward?

Hi @timhoffm, Hi @scottshambaugh :-
I've come up with a draft PR #28841 for the Issue #28408 ('adjustable mplot3d mouse rotation style'); it incorporates most of the changes proposed in the current PR ('fix 3D rotation precession'). It lets you choose what mouse rotation style you want (the traditional style; or an arcball, with a trade-off: with natural speed, but you pay the price: precession; or an arcball without precession, but you pay the price: double the angular rate; or a trackball).

It also makes the trackball/arcball size an rcparameter; originally, the size was hard-coded, and set a tad large; and then this PR actually grew it even larger (rather than shrinking, as I mistakenly thought initially). (This made the arcball almost work as a trackball.)

What is still absent is the introduction of an `np.clip()`, I have hesitations about that since I don't understand why it is needed (maybe it is something for the final-merge PR, if it really unavoidable).

In any case, could you take a look and let me know what you think about it?

timhoffm commented:
Sorry, I'm out of this discussion. [As stated](https://github.com/matplotlib/matplotlib/issues/28408#issuecomment-2172591569) I'm a rare and naive 3D user and want the rotation to "just work". From that perspective (which I assume is also that of many users) I would argue that configurability is not necessary and we could have just one fixed algorithm. There may be aspects that I overlook, which make configuration neccessary. But that's that's exactly the point why I leave the discussion to you experts.

MischaMegens2 commented:
Hi @scottshambaugh,

It's up to us then now, it looks like...
;)

scottshambaugh commented:
After reviewing https://github.com/matplotlib/matplotlib/pull/28841, I think it would be most preferable for that to go in instead of this one. But if it's not ready before the v3.10 merge point, I think this PR should go in as a stop-gap to improve the default behavior in the new release.

scottshambaugh commented:
With https://github.com/matplotlib/matplotlib/pull/28841 merged in, this PR is no longer necessary. Closing.
